elastic/fleet-server
# File delivery

## What is the problem this PR solves?

Delivering user-uploaded files (via kibana) to integrations (specifically, endpoint use-case)

## How does this PR solve the problem?

an API route was added, `/api/fleet/file/[ID]`.  Authenticated calls to this will locate file metadata documents in elasticsearch for the presence of a file, and that the authenticated agent is present in the file's `file.Meta.target_agents` field (i.e. only allowed client IDs can download). Fleet server iterates through the file contents to send on the response pipe

## How to test this PR locally

- Create destination indices in Elasticsearch matching these templates (until the PR is merged): https://github.com/elastic/elasticsearch/pull/96504
- Use a recent Dev Endpoint that includes file-requesting feature
- In kibana, create a send-file action to send a file to an Endpoint

Alternatively, a simpler setup would be to check out this PR, and modify the authentication section in `internal/pkg/api/handleFileDelivery.go` line 48. Comment out the `authAgent` call (or otherwise mock it), and put a different, perhaps static ID as the last parameter to the `FindFileForAgent` call a few lines down  (e.g. `... fileID, "mockID")`). As long as a file is uploaded to the right indices (`.fleet-filedelivery-meta-endpoint` and `.fleet-filedelivery-data-endpoint`) with that matching agent_id in the `target_agents` field, then a command-line curl, wget or similar to `https://localhost:8221/api/fleet/file/[ID]` should retrieve the file like any other file-based HTTP url.

## Design Checklist




## Checklist





## Related issues
